x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/audit/logfile/LoggingAuditTrail.java

```diff
@@ -727,16 +814,313 @@ public void runAsDenied(String requestId, Authentication authentication, RestReq
+            } else {
+                logEntry.with(EVENT_ACTION_FIELD_NAME, "disable_user");
+                logEntry.with(DISABLE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            }
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(PutPrivilegesRequest putPrivilegesRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "put_privileges");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject()
+                    // toXContent of {@code ApplicationPrivilegeDescriptor} does a good job
+                    .field("privileges", putPrivilegesRequest.getPrivileges())
+                    .endObject();
+            logEntry.with(PUT_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(CreateApiKeyRequest createApiKeyRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "create_apikey");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject();
+            withRequestBody(builder, createApiKeyRequest);
+            builder.endObject();
+            logEntry.with(CREATE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        LogEntryBuilder withRequestBody(GrantApiKeyRequest grantApiKeyRequest) throws IOException {
+            logEntry.with(EVENT_ACTION_FIELD_NAME, "create_apikey");
+            XContentBuilder builder = JsonXContent.contentBuilder().humanReadable(true);
+            builder.startObject();
+            withRequestBody(builder, grantApiKeyRequest.getApiKeyRequest());
+            GrantApiKeyRequest.Grant grant = grantApiKeyRequest.getGrant();
+            builder.startObject("grant")
+                        .field("type", grant.getType());
+                        if (grant.getUsername() != null) {
+                            builder.startObject("user")
+                                    .field("name", grant.getUsername())
+                                    .field("has_password", grant.getPassword() != null)
+                                    .endObject(); // user
+                        }
+                        if (grant.getAccessToken() != null) {
+                            builder.field("has_access_token", grant.getAccessToken() != null);
+                        }
+                    builder.endObject(); // grant
+            builder.endObject();
+            logEntry.with(CREATE_CONFIG_FIELD_NAME, Strings.toString(builder));
+            return this;
+        }
+
+        private void withRequestBody(XContentBuilder builder, CreateApiKeyRequest createApiKeyRequest) throws IOException {
+            TimeValue expiration = createApiKeyRequest.getExpiration();
+            builder.startObject("apikey")
```

Comment:
Often times (if not all), in serialisation form, it is called `api_key`.

Reply:
It shows `apikey` in existing audit logs already. Your call, should the new events use `api_key` or `apikey`?

Reply:
That's unfortunate... Since we already have `apikey` in the audit log, let's just keep it this way.